### ISSUE TITLE
How to validate the trained weights?

### ISSUE THREAD

**Issue by MheadHero** [sovit-123/fastercnn-pytorch-training-pipeline#10]

**Comment by sovit-123** [comment]
Hi @MheadHero 
You can either take the final validation mAP from the final epoch or else another way is to run a test script over a test set if you have one with ground truth for your dataset. Now, I am well aware that the test script is not yet present in the repo and that's what my next focus will be to upload in the next few days.

**Comment by MheadHero** [comment]
Hi sir, can you share some idea on how to write a test script to validate on validation set and testing set like the val.py in yolov5? Because currently I also following a repo to implement YOLOv4 but it didn't provide test/val.py script too. Moreover, that repo has been inactive for long time, so I can't find people to ask. But my lecturer ask me to write the script myself and I look around at several test.py and train.py available on github, I also have no idea, how to write it. So can you share a bit? Sorry for all the troublesome but you are the only professional I find online doing similar work and is willing to reply me.

**Comment by sovit-123** [comment]
Hi. 
If you already have a test set with the images and labels in Pascal VOC format, then it should be pretty straightforward. The simplest way to get started will be to replicate the [`evaluate()`](https://github.com/sovit-123/fastercnn-pytorch-training-pipeline/blob/2e3d2cbd33e73be236ed004eea9afb9670b6fd35/torch_utils/engine.py#L95) function in `engine.py` which is available inside the `torch_utils` directory. In other words, the test.py will just run validation on an unseen test set. I think that should get you started.

**Comment by sovit-123** [comment]
@MheadHero Will do it soon. As of now, there are only two contributors, so things are a bit slow. But will try to include in the next push.

**Comment by sovit-123** [comment]
@MheadHero Sorry for the delay!
I have added the `validate.py` script to either validate on test data (unseen during training) or on the validation data (used in the validation loop during training). A few simple things to take care of.

- Suppose that your data YAML file is like the following:
 ```
# Images and labels direcotry should be relative to train.py
TRAIN_DIR_IMAGES: '../input/Aquarium Combined.v2-raw-1024.voc/train'
TRAIN_DIR_LABELS: '../input/Aquarium Combined.v2-raw-1024.voc/train'
VALID_DIR_IMAGES: '../input/Aquarium Combined.v2-raw-1024.voc/valid'
VALID_DIR_LABELS: '../input/Aquarium Combined.v2-raw-1024.voc/valid'

# Class names.
CLASSES: [
    '__background__',
    'fish', 'jellyfish', 'penguin',
    'shark', 'puffin', 'stingray',
    'starfish'
]

# Number of classes (object classes + 1 for background class in Faster RCNN).
NC: 8

# Whether to save the predictions of the validation set while training.
SAVE_VALID_PREDICTION_IMAGES: True
```
The above does not contain `TEST_DIR_IMAGE` and `TEST_DIR_LABELS`. So, during validation, `VALID_DIR_IMAGES` and `VALID_DIR_LABELS` will be used.
- Your training command may look like the following:
`python train.py --model fasterrcnn_mobilenetv3_large_320_fpn --config data_configs/aquarium.yaml --epochs 10 --project-name aquarium --batch-size 4 --no-mosaic`
- Then your validation command will become:
`python validate.py --config data_configs/aquarium.yaml --model fasterrcnn_mobilenetv3_large_320_fpn --weights outputs/training/aquarium/last_model_state.pth`

I hope this helps. If this solves your issue, consider leaving a comment and a reaction and closing it as well. This may help others in the future as well.

**Comment by MheadHero** [comment]
First of all, thank you so much Sir! You will definitely be one of the tech master in the future! 
![image](https://user-images.githubusercontent.com/24875124/170301218-a92bf4e3-1328-4181-8c91-0385753964c6.png)
I have 2 questions, first question is from the img above, means that if my yaml file consist of TEST_DIR_IMAGES and TEST_DIR_LABELS, then it will load my testing set right? 

Second question is a newbie question, actually why when training, you guys didn't show the train map?(what if I want see the train map?) but directly show validated map? (I saw in train.py when evaluate, it feed in the validation dataset right)

**Comment by sovit-123** [comment]
@MheadHero 
Yes, if you have test images that are not used in the `train.py` script, they will be used in the `validate.py` instead of the `VALID_DIR_IMAGES` and corresponding labels.

The answer to the second question is situation-dependent. It is just the norm to show loss during training and mAP for the validation loop only. I think the main reason is that for learning the training loop and its loss is used. And we don't care much about the training data mAP. Instead we want to know the mAP for either the validation data or the test data. In other words, we want the mAP for the data that is not used in training so that we can evaluate our model based on that. I think that's one fo the reasons.
The same approrach is taken by the YOLOv5 repo as well. So, I thought, its just better to follow the standard.
